8238812: assert(false) failed: bad AD file

r-v-raghav opened the pull request:
Found that the earlier PR bug analysis and fix details done is wrong.
Request help to review this revised fix proposal.

Now understood need to fix two types of issues related to the bug-

(i). Crash with only `CmpI` nodes case.
Sample test case -
```
public static void test2() {
  for (int i = 5; i > -20; i -= 5) {
    switch (i) {
      case 0:
      case 10:
      case 20:
      case 30:
      case 40:
      case 50:
      case 60:
      case 100:
        res = 42;
        break;
    }
  }
}
```

// This generates the following subgraph:
```
/*
// i: -20..0
if (i != 0) {
  // i: -20..-1
  if (i < 0) {   <- /*This is always true but not folded by C2*/
    // Fall through
  } else {
    ...
    CastII(i-1, 0..4) <- /*Replaced by TOP because i-1 range is -21..-1 but still considered reachable by C2 although it is dead code*/
    ...
  }
} else {
  StoreI   <- Due to this additional store on, IfNode::has_shared_region returns false and the fold compares optimization does not kick in
}
*/
```

Sample IR extracts -
```
 77  Phi  ===  399  25  389  [[ 658  664  389  80  662  660  659 ]]
 25  ConI  ===  0  [[ 80  77 ]]  #int:0
 80  CmpI  === _  77  25  [[ 81  89 ]]
 81  Bool  === _  80  [[ 82 ]] [ne]
 82  If  ===  399  81  [[ 85  87 ]]
 85  IfFalse  ===  82  [[ 102  336 ]]
 87  IfTrue  ===  82  [[ 90 ]] #1
 89  Bool  === _  80  [[ 90 ]] [lt]
 90  If  ===  87  89  [[ 91  92 ]]
 91  IfTrue  ===  90  [[ 102 ]] #1
 92  IfFalse  ===  90  [[ 156 ]] #0
 336  StoreI  ===  85  359  335  332  [[ 337 ]]
 156  Jump  ===  92  1  [[ 157  160  .......
```

Fix proposal here includes :
- Extract the existing non-`CmpU` related optimization out of old `fold_compares_helper()`,
- form new method
(new `fold_dominated_if()` added and renamed old `fold_compares_helper` as `fold_to_unsigned()`).
- Call it properly from `fold_compares()` and related code adjustments.




(ii). `fold_compares()` not handling `CmpU` nodes cases.

Sample basic test case -
```
public static void test3() {
  int local = 0;
  for (int i = 5; i > -20; i -= 5) {
    switch (i) {
    case 0:
    case 10:
    case 20:
    case 30:
    case 40:
    case 50:
    case 100:
      local = 42;
      break;
    }
  }
}
```

```
// This generates the following subgraph:
/*
// i: -20..0
if (i != 0) {
  // i: -20..-1
  if (i u< 101) {    <- /*This is always false but not folded by C2 because CmpU is not handled*/
    CastII(i-1, 0..49)  <- /*Replaced by TOP because i-1 range is -21..-1 but still considered reachable by C2 although it is dead code*/
  } else {
    ...
  }
} else {
  ...
}
*/
```

Sample IR extracts -
```
 77  Phi  ===  363  20  353  [[ 573  577  572  80  575  353  345  574 ]]
 20  ConI  ===  0  [[ 80  77 ]]  #int:0
 80  CmpI  === _  77  20  [[ 81 ]]
 81  Bool  === _  80  [[ 82 ]] [ne]
 82  If  ===  363  81  [[ 85  87 ]]
 85  IfFalse  ===  82  [[ 102 ]] #0
 87  IfTrue  ===  82  [[ 337 ]] #1
 95  ConI  ===  0  [[ 345 ]]  #int:101
 345  CmpU  === _  77  95  [[ 346 ]]
 346  Bool  === _  345  [[ 337 ]] [lt]
 337  If  ===  87  346  [[ 338  339 ]]
 338  IfFalse  ===  337  [[ 102 ]] #0
 339  IfTrue  ===  337  [[ 135 ]] #1
 135  Jump  ===  339  1  [[ 136  139  142  .....
```

Proposed fix includes:
Enabling `Op_CmpU` support in -
- `cmpi_folds()`,
- `filtered_int_type()` - returns possible restrictive type for input value based on condition control flow for the if (supported `CmpU` for various condition control types).
- `fold_dominated_if()` - checks if dominating if determines the result of current if and above example test case is a basic variant possibility where both `CmpNode`s compare a common value directly. (`Node 77 Phi` in above sample). But as mentioned in the comments of new helper function `get_base_comparing_value()`, there are other possibilities where there might be an `AddINode` with a constant increment present in between the `CmpNode`s and the common value compared. 
(Sample IR extract with `AddI` in between `CmpNode`s :  )
```
 76  Phi  ===  364  79  354  [[ 576  580  575  80  578  354  344  577 ]]  #int:0..20:www
 79  ConI  ===  0  [[ 80  76 ]]  #int:20
 80  CmpI  === _  76  79  [[ 81 ]]
 81  Bool  === _  80  [[ 82 ]] [ne] 
 82  If  ===  364  81  [[ 85  87 ]]
 85  IfFalse  ===  82  [[ 102 ]] #0
 87  IfTrue  ===  82  [[ 338 ]] #1
 343  ConI  ===  0  [[ 344 ]]  #int:-20
 344  AddI  === _  76  343  [[ 348 ]] 
 347  ConI  ===  0  [[ 348 ]]  #int:281
 348  CmpU  === _  344  347  [[ 349 ]] 
 349  Bool  === _  348  [[ 338 ]] [lt]
 338  If  ===  87  349  [[ 339  340 ]]
 339  IfFalse  ===  338  [[ 102 ]] #0 
 340  IfTrue  ===  338  [[ 136 ]] #1
 136  Jump  ===  340  1  [[ 137  140  143  146 .........
```
All these variants are handled in `get_base_comparing_value()` and related test cases added.

- and related changes in `fold_compares()`.


Found no issues with this proposed fix for various unit tests, reported java_fuzzer case, tier1-7 and pre-integeration tests

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8238812](https://bugs.openjdk.java.net/browse/JDK-8238812): assert(false) failed: bad AD file


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2758/head:pull/2758`
`$ git checkout pull/2758`

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8238812](https://bugs.openjdk.java.net/browse/JDK-8238812): assert(false) failed: bad AD file


### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Roland Westrelin](https://openjdk.java.net/census#roland) (@rwestrel - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2758/head:pull/2758`
`$ git checkout pull/2758`

TobiHartmann commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -837,8 +857,102 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
+    const TypeInt* val_t = igvn->type(this_val->in(2))->isa_int();
+    if (val_t && val_t->is_con()) {
+      this_adj_val = val_t->_lo;
+      res_val = this_val->in(1);
```
Could be replaced by `res_val = dom_val;`

TobiHartmann commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -837,8 +857,102 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
+    res_val = this_val;
+  } else if (this_val->is_Add() && this_val->in(1) && this_val->in(1) == dom_val) {
+    const TypeInt* val_t = igvn->type(this_val->in(2))->isa_int();
+    if (val_t && val_t->is_con()) {
```
`val_t` implicitly casts a pointer to a boolean, should be replaced by `val_t != NULL`. Same below.

TobiHartmann commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -837,8 +857,102 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
+  jint dom_adj_val = 0;
+  Node* n = NULL;
+
+  n = get_base_comparing_value(dom_if, igvn, this_adj_val, dom_adj_val);
```
No need to initialize `n` above. Should be `Node* n = get_base_comparing_value(...`

TobiHartmann replied:
Actually, you are not even using `n` and can therefore just return a boolean from `get_base_comparing_value` and simplify that method accordingly.

TobiHartmann commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -837,8 +857,102 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
+
+// Check if dominating if determines the result of this if
+bool IfNode::fold_dominated_if(ProjNode* proj, PhaseIterGVN* igvn) {
+  Node* this_cmp = in(1)->in(1);
```
We only ever use `this_cmp->in(1)` so you can replace this by `Node* this_val = in(1)->in(1)->in(1)`. Same for `dom_cmp` below.

TobiHartmann commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -837,8 +857,102 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
+      failtype = dom_cmp->in(1)->as_Add()->add_ring(failtype, TypeInt::make(-dom_adj_val))->is_int();
+    }
+    for (int i = 0; i < 2; ++i) {
+      const TypeInt* type2 = filtered_int_type(igvn, this_cmp->in(1), proj_out(i));
```
Why not just name this `type`?

TobiHartmann commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -837,8 +857,102 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
+          assert(this_cmp->in(1)->is_Add() != false, "sanity");
+          type2 = this_cmp->in(1)->as_Add()->add_ring(type2, TypeInt::make(-this_adj_val))->is_int();
+        }
+        const TypeInt* res_type = failtype->join(type2)->is_int();
```
No need to declare `res_type`, just use `type`.

chhagedorn commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -700,16 +720,17 @@ const TypeInt* IfNode::filtered_int_type(PhaseGVN* gvn, Node* val, Node* if_proj
 
 // Is the comparison for this If suitable for folding?
-bool IfNode::cmpi_folds(PhaseIterGVN* igvn, bool fold_ne) {
+bool IfNode::cmpi_folds(PhaseIterGVN* igvn) {
```
We should probably rename it into `cmp_folds` now that it is also for unsigned and also update the description above. Shouldn't the description be at `fold_compares()` anyways? But maybe we want to clean this up in a follow-up RFE.

r-v-raghav replied:
Yes, will rename `cmpi_folds` to `cmp_folds`.

Not changing the fold compares description location for now.
For now maybe it is okay location for description as that is starting of fold_compares implementation (all methods above fold_compares definition seems mainly its helper functions)
.e.g.: See the comment and declarations in [cfgnode.hpp]
........
private:
  // Helper methods for fold_compares
  bool cmp_folds(PhaseIterGVN* igvn);
  bool is_ctrl_folds(Node* ctrl, PhaseIterGVN* igvn);
Or yes agree we can change this in a follow-up RFE.

chhagedorn replied:
Yeah, it's probably best to leave it as it is at the moment and clean it up later in an RFE.

r-v-raghav commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -837,8 +857,94 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
   return false;
 }
 
+// There might be an AddINode (marked with *) with a constant increment
```
Never found a SubNode in this case. Always AddI node with integer constant as second input (this constant 
can be positive or negative values)

TobiHartmann replied:
That's because we always convert `SubINodes` with constant second operand to `AddINodes`:
https://github.com/openjdk/jdk/blob/3d3eb5c8d31ecd80309fdcbb67cd2aeeb2037c52/src/hotspot/share/opto/subnode.cpp#L181

chhagedorn replied:
Okay I see, thanks for the explanation.

r-v-raghav commented:
Thanks to Tobias and Christian for all help review comments, suggestions.
Please I could address all the comments suggestions above so far.

TobiHartmann reviewed (approved):
This looks good to me.

chhagedorn reviewed (approved):
Looks good to me!

rwestrel commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -628,31 +630,51 @@ const TypeInt* IfNode::filtered_int_type(PhaseGVN* gvn, Node* val, Node* if_proj
               if (hi != min_jint) {
                 hi = hi - 1;
               }
               break;
```
I think there's a bug here (case BoolTest::lt, unsigned comparison) . if cmp2 is [-1, 1] for instance, then lo and hi are unchanged. If the type of val is TypeInt::INT then the resulting type once join is applied is the interval [-1, 1]. But for instance 42 <u -1, 42 is in TypeInt::INT, -1 is in [-1, 1] but 42 is not in the resulting interval. I wonder if other cases in this method has similar issues. Can you fix that one and verify others?

rwestrel commented:
This fixes the bug when the tests that need to be optimized out are next to one an another and the logic for folding comparisons can be used. It's hard to be certain that it's not possible for the 2 tests be separated by some other control flow. In that case the fix would fail.

A conservative fix would be to not use a narrow type in the CastII that tableswitch/lookupswitch switch introduces. It would be a simpler fix and a more robust one.

r-v-raghav commented:
Thank you @rwestrel for the comments. 
Understood your point.
Based on the review comments now planning to -

i. Do proposal by Roland and remove the narrow type from the CastII as fix for this bug task.
As of now found no issues with reported java_fuzzer cases / unit tests, mach5 tier 1-7 & pre-integration tests
for the following changes to avoid use any narrow type in the `CastII` introduced by `lookupswitch`.

[src/hotspot/share/opto/parse2.cpp]
```
- const TypeInt* ikeytype = TypeInt::make(0, num_cases, Type::WidenMin);
............
- key_val = C->conv_I2X_index(&_gvn, key_val, ikeytype, control());
..............
+#ifdef _LP64
+ key_val = C->constrained_convI2L(&_gvn, key_val, TypeInt::INT, control());
+#endif
```


ii. Integrate the other proposed fix to improve `fold_compares` optimizations as a separate enhancement.
(after addressing the comments in `filtered_int_type()`)

r-v-raghav commented:
Please note as per the above comments now updated the fix. Request help with review. Thanks.

TobiHartmann reviewed (approved):
This looks good to me. Please file a follow-up enhancement to integrate the optimization, I think it's still valuable.

TobiHartmann reviewed (changes requested):
After quickly discussing this with @rwestrel, it turned out that you need to set `CastIINode::_carry_dependency` to prevent the `CastIINode` with `TypeInt::INT` from being eliminated and re-introducing ([JDK-6675699](https://bugs.openjdk.java.net/browse/JDK-6675699)).

r-v-raghav commented:
Understood that there might still be a problem with latest changes - 
A `CastII` with `TypeInt::Int` can be removed by IGVN if `_carry_dependency` is not set.
Working on revised fix.

rwestrel reviewed (approved):
Thanks for updating the change.

TobiHartmann reviewed (approved):
Looks good to me too.

chhagedorn reviewed (approved):
Looks good!
